test/optim/test_optim.py

```diff
@@ -834,6 +835,49 @@ def test_rprop(self):
             )
 
 
+    def test_lbfgs(self):
+        self._test_basic_cases(
+            lambda weight, bias: LBFGS([weight, bias]), ignore_multidevice=True
+        )
+        self._test_basic_cases(
+            lambda weight, bias: LBFGS(
+                [weight, bias], line_search_fn="strong_wolfe"
+            ),
+            ignore_multidevice=True,
+        )
+
+
+    def test_lbfgs_complex(self):
+        for linesearch in ("strong_wolfe", None):
+            complex_params = [torch.randn(10, 5, dtype=torch.complex64, requires_grad=True), torch.randn(10, dtype=torch.complex64, requires_grad=True), torch.randn(10, 5, dtype=torch.float32, requires_grad=True)]
```

Comment:
is the last param being float32 intentional?

Reply:
Yes. 
We would like to use bfgs with both complex and real parameters. Our first try to fix the issue did not take this into account and it failed when real and complex gradients were concatenated in _gather_flat_grad.